torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -0,0 +1,281 @@
+          "MPI_THREAD_SERIALIZED. This is required by "
+          "c10d package");
+    }
+    if (std::atexit(mpiExit)) {
+      throw std::runtime_error("Fail to register the MPI exit handler");
+    }
+    mpiInitialized_ = true;
+  }
+
+  // Update the world size and rank
+  MPI_CHECK(MPI_Comm_size(MPI_COMM_WORLD, &size_));
+  MPI_CHECK(MPI_Comm_rank(MPI_COMM_WORLD, &rank_));
+
+  if (mpiThreadSupport_ != MPI_THREAD_MULTIPLE && numProcessGroups_ >= 1) {
+    throw std::runtime_error(
+        "More than one process group created, "
+        "this is not supported due to the used MPI "
+        "implementation doesn't provide the full support "
+        "of multi-threading");
+  }
+  // increase the total PG count
+  ++numProcessGroups_;
+
+  // Start the worker thread accepting MPI calls
+  workerThread_ = std::thread(&ProcessGroupMPI::runLoop, this);
+}
+
+ProcessGroupMPI::~ProcessGroupMPI() {
+  destroy();
+}
+
+void ProcessGroupMPI::destroy() {
+  std::unique_lock<std::mutex> lock(pgMutex_);
+
+  while (!queue_.empty()) {
+    queueConsumeCV_.wait(lock);
+  }
+  // Queue is empty, signal stop
+  stop_ = true;
+
+  // Release lock to allow threads to terminate
+  queueProduceCV_.notify_all();
+
+  lock.unlock();
+
+  // Join the single worker thread
+  workerThread_.join();
+
+  // Decrease the number of PG created
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  --numProcessGroups_;
+}
+
+void ProcessGroupMPI::abort() {
+  destroy();
+  MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
+}
+
+void ProcessGroupMPI::runLoop() {
+  std::unique_lock<std::mutex> lock(pgMutex_);
+
+  while (!stop_) {
+    if (queue_.empty()) {
+      queueProduceCV_.wait(lock);
+      continue;
+    }
+
+    auto workTuple = std::move(queue_.front());
+
+    queue_.pop_front();
+    queueConsumeCV_.notify_one();
+
+    auto& workEntry = std::get<0>(workTuple);
+    auto& work = std::get<1>(workTuple);
+
+    lock.unlock();
+
+    try {
+      workEntry->run(workEntry);
+      work->finish();
+    } catch (...) {
+      work->finishWithException(std::current_exception());
+    }
+
+    lock.lock();
+  }
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::enqueue(
+    std::unique_ptr<WorkEntry> entry) {
+  auto work = std::make_shared<WorkMPI>();
+  std::unique_lock<std::mutex> lock(pgMutex_);
+  queue_.push_back(std::make_tuple(std::move(entry), work));
+  queueProduceCV_.notify_one();
+  return work;
+}
+
+void ProcessGroupMPI::checkSingleTensor(
+    const std::vector<at::Tensor>& tensors) {
+  if (tensors.size() != 1) {
+    throw std::runtime_error(
+        "MPI process group only supports a single "
+        "tensor op");
+  }
+  if (!tensors[0].is_contiguous()) {
+    throw std::runtime_error("input tensor has to be contiguous");
+  }
+  if (tensors[0].is_cuda() && !cudaAwareMpiCheck()) {
+    throw std::runtime_error(
+        "CUDA tensor detected and the MPI used doesn't "
+        "have CUDA-aware MPI support");
+  }
+}
```

Comment:
Can be a static function (like the asserts in Utils.hpp, but local to this file since it includes an MPI check)